dot_config/nvim/lua/_lsp.lua

```diff
@@ -0,0 +1,249 @@
+    vim.cmd [[augroup lsp_formatting]]
+    vim.cmd [[autocmd!]]
+    vim.cmd [[autocmd BufWritePre <buffer> :lua vim.lsp.buf.formatting_sync({}, 5000)]]
+    vim.cmd [[augroup END]]
+
+    buf_set_keymap("n", "<space>f", "<cmd>lua vim.lsp.buf.formatting()<CR>", opts)
+  elseif client.resolved_capabilities.document_range_formatting then
+    buf_set_keymap("n", "<space>f", "<cmd>lua vim.lsp.buf.range_formatting()<CR>", opts)
+  end
+end
+
+-- Stolen from lspconfig/tsserver, would have been nice to be able to import
+local get_ts_root_dir = function(fname)
+  return util.root_pattern("tsconfig.json")(fname) or
+  util.root_pattern("package.json", "jsconfig.json", ".git")(fname);
+end
+
+-- Make tsserver work with yarn v2
+local get_tsserver_exec = function()
+  local ts_root_dir = get_ts_root_dir(vim.fn.getcwd())
+  local coc_settings = ts_root_dir .. "/.vim/coc-settings.json"
+
+  -- not yarn v2 project
+  if util.path.exists(coc_settings) == false then
+    return "node_modules/.bin/tsserver"
+  else
+    local file = io.open(coc_settings):read("*a")
+    local coc_json = vim.fn.json_decode(file)
+    local ts_key = "tsserver.tsdk"
+    return coc_json[ts_key] .. "/tsserver.js"
+  end
+end
+
+local on_publish_diagnostics = function(prefix)
+  return function(err, method, params, client_id, bufnr, config)
+    vim.tbl_map(
+      function(value)
+        value.message = prefix .. value.message
+      end,
+      params.diagnostics
+    )
+
+    return vim.lsp.with(
+      vim.lsp.diagnostic.on_publish_diagnostics, {
+        -- show underline
+        underline = true,
+        -- Enable signs
+        signs = {
+          -- Make priority higher than vim-signify
+          priority = 100
+        },
+        -- Disable virtual_text
+        virtual_text = false,
+        -- show diagnostics on exit from insert
+        update_in_insert = true
+      }
+    )(err, method, params, client_id, bufnr, config)
+  end
+end
+
+local tsserver_exec = get_tsserver_exec()
+nvim_lsp.tsserver.setup {
+  handlers = {
+    ["textDocument/publishDiagnostics"] = on_publish_diagnostics("[tsserver] ")
+  },
+  cmd = {
+    "typescript-language-server",
+    "--stdio",
+    "--tsserver-path",
+    tsserver_exec
+  },
+  on_attach = function(client, bufnr)
+    client.resolved_capabilities.document_formatting = false
+    on_attach(client, bufnr)
+  end,
+  capabilities = lsp_status.capabilities
+}
+
+local map_table_to_key = function(tbl, key)
+  return vim.tbl_map(function(value)
+    return value[key]
+  end, tbl)
+end
+
+local diagnosticls_languages = {
+  javascript = {
+    linters = { "eslint" },
+    formatters = { "prettier" }
+  },
+  javascriptreact = {
+    linters = { "eslint" },
+    formatters = { "prettier" }
+  },
+  typescript = {
+    linters = { "eslint" },
+    formatters = { "prettier" }
+  },
+  typescriptreact = {
+    linters = { "eslint" },
+    formatters = { "eslint", "prettier" }
+  },
+}
+require'lspconfig'.diagnosticls.setup{
```

Comment:
Found [diagnosticls](https://github.com/iamcco/diagnostic-languageserver) had better performance and features than [efm](https://github.com/mattn/efm-langserver). Feature wise I don't think it wouldn't report the severity or location very well (just gave the line rather than line & range)